jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java

```diff
@@ -0,0 +1,197 @@
+package com.jtelegram.api.menu.impl;
+
+import com.jtelegram.api.TelegramBot;
+import com.jtelegram.api.chat.Chat;
+import com.jtelegram.api.chat.id.ChatId;
+import com.jtelegram.api.events.inline.keyboard.CallbackQueryEvent;
+import com.jtelegram.api.ex.TelegramException;
+import com.jtelegram.api.inline.keyboard.InlineKeyboardButton;
+import com.jtelegram.api.menu.Menu;
+import com.jtelegram.api.menu.MenuButton;
+import com.jtelegram.api.menu.MenuButtonResponse;
+import com.jtelegram.api.menu.MenuContext;
+import com.jtelegram.api.menu.MenuState;
+import com.jtelegram.api.menu.OnClickHandler;
+import com.jtelegram.api.message.impl.TextMessage;
+import com.jtelegram.api.requests.inline.AnswerCallbackQuery;
+import com.jtelegram.api.requests.message.edit.EditTextMessage;
+import com.jtelegram.api.requests.message.framework.ParseMode;
+import com.jtelegram.api.requests.message.send.SendText;
+import com.jtelegram.api.user.User;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class MenuImpl implements Menu {
+
+    private static final AtomicInteger menuIdCounter = new AtomicInteger(0);
```

Comment:
Menu IDs here are generated on the fly, increment from zero, but the result of this is that when the bot restarts, old menus will make execute callbacks to newly generated ones as they share the same id from different states.

The solution for this is to create menus with string identifiers provided by the developer (A-z,0-9), and allow menus to have multiple viewers (which I think you allow here). So instead of your callback data being `menuId stateId row button` it should be `menuId chatId stateId row button`. An additional consideration would be how you allow developers to add metadata to the `MenuContext` class, since that isn't persistent either. Removing this ability would force the developer to implement their own chat -> context appropriately for the menu/bot they are building, which would be more functional than forgetting everything on restart.

If you want, you could implement a saving system for `MenuContext` and take care of it for the developer, providing them with different options, but I think that goes beyond the reach of this API.